File: winter2009/comp2150/a5/q2/lib/cashier_arrival_event.rb
# file: lib/cashier_arrival_event.rb

# Models the arrival of a cashier into a simulation of a supermarket.
class CashierArrivalEvent < ArrivalEvent

  def initialize(time, eid)
    super(time)
    @cashier = Cashier.new(time, eid)
  end

  # Event action. Set up camp at the checkout.
  def run
    SoopaStore.checkout.cashier = @cashier
    @cashier.checkout = SoopaStore.checkout
    SoopaStore.cashiers.add @cashier
    puts "Time #{SoopaStore.current_time}  : Cashier #{@cashier.eid} arrives & starts work"
  end
  
end

File: winter2009/comp2150/a3/q1/CashierArrivalEvent.java
/**
 * Models the arrival of an employee in a simulation
 *
 * <pre>
 * CashierArrivalEvent: 
 *   (open first available till)
 *   T self -> ShiftCompletedEvent @ quittingTime
 * </pre>
 */
public class CashierArrivalEvent extends ArrivalEvent {

  Checkout checkout; ///< The specific Checkout involved in this event.
  
  /**
   * Construct a new CashierArrivalEvent
   *
   * @param arrival time at which employee arrives
   * @param employee employee arriving
   * @externals Cashier
   */
  public CashierArrivalEvent(int arrival, Cashier employee) {
    super(arrival, employee);
  }

  /**
   * method called when this event "happens". Find the first available
   * checkout, then trigger a ShiftCompletedEvent at quittingTime.
   * @externals Cashier, ShiftCompletedEvent
   */
  public void action() {
    SoopaStore.readNextArrival();
    Cashier employee = (Cashier)this.getActor();
    checkout = SoopaStore.firstUnmannedCheckout();
    checkout.setCashier(employee);
    checkout.setOpen(true);
    int quittingTime = employee.getQuittingTime();
    Event event = new ShiftCompletedEvent(quittingTime, employee, checkout);
    SoopaStore.getEventList().insertInOrder(event);
    
    System.out.printf("Time %d: Cashier %d arrives (Speed %d, quitting time %d, checkout %d)\n",
                      time,
                      employee.getId(),
                      employee.getSpeed(),
                      employee.getQuittingTime(),
                      checkout.getIndex());
  }
  
}